src/main/scala/nl/knaw/dans/easy/multideposit/Action.scala

```diff
@@ -15,32 +15,155 @@
  */
 package nl.knaw.dans.easy.multideposit
 
-import scala.util.{Success, Try}
+import scala.util.{ Failure, Success, Try }
+import nl.knaw.dans.lib.error.{ CompositeException, TraversableTryExtensions }
+import nl.knaw.dans.lib.logging.DebugEnhancedLogging
+
+import scala.collection.mutable
+import scala.util.control.NonFatal
 
 /**
-  * An action to be performed by Process SIP. It provides three methods that can be invoked to verify
-  * the feasibility of the action, to perform the action and - if necessary - to roll back the action.
-  */
-trait Action {
+ * An action to be performed by Process SIP. It provides three methods that can be invoked to verify
+ * the feasibility of the action, to perform the action and - if necessary - to roll back the action.
+ */
+/*
+  To future developers: this class is a semigroup. It satisfies the associativity law as defined
+  in https://wiki.haskell.org/Typeclassopedia#Laws_4.
+ */
+trait Action extends DebugEnhancedLogging {
+
+  private def logPreconditions(): Unit = {
+    logger.info(s"Checking preconditions of ${getClass.getSimpleName} ...")
+  }
+  private def logExecute(): Unit = {
+    logger.info(s"Executing action of ${getClass.getSimpleName} ...")
+  }
+  private def logRollback(): Unit = {
+    logger.info(s"An error occurred. Rolling back action ${getClass.getSimpleName} ...")
+  }
+  private[Action] def innerCheckPreconditions: Try[Unit] = Try(logPreconditions()).flatMap(_ => checkPreconditions)
+  private[Action] def innerExecute(): Try[Unit] = Try(logExecute()).flatMap(_ => execute())
+  private[Action] def innerRollback(): Try[Unit] = Try(logRollback()).flatMap(_ => rollback())
+
+  /**
+   * Verifies whether all preconditions are met for this specific action.
+   *
+   * @return `Success` when all preconditions are met, `Failure` otherwise
+   */
+  protected def checkPreconditions: Try[Unit] = Success(())
```

Comment:
just Scala convention: `checkPreconditions` is a _pure_ function that does not modify any data, do any I/O or anything like that (it has no side-effects). `execute` and `rollback` however are expected to be _unpure_ functions, because they are supposed to write data to files, create folders on the filesystem and undo all that stuff.

See also http://stackoverflow.com/questions/3877953/why-to-use-empty-parentheses-in-scala-if-we-can-just-use-no-parentheses-to-defin